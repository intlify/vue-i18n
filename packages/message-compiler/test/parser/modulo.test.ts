import {
  createParser,
  NodeTypes,
  PluralNode,
  MessageNode
} from '../../src/parser'

let spy: any // eslint-disable-line @typescript-eslint/no-explicit-any
beforeEach(() => {
  spy = jest.fn()
})

test('hi %{name}!', () => {
  const text = `hi %{name}!`
  const parser = createParser({ onError: spy })
  const ast = parser.parse(text)

  expect(ast).toMatchSnapshot()
  expect(spy).not.toHaveBeenCalled()

  expect(ast.type).toEqual(NodeTypes.Resource)
  expect(ast.body.type).toEqual(NodeTypes.Message)
  const message = ast.body as MessageNode
  expect(message.items).toHaveLength(3)
  expect(message.items).toMatchObject([
    {
      type: NodeTypes.Text,
      value: 'hi '
    },
    {
      type: NodeTypes.Named,
      key: 'name'
    },
    {
      type: NodeTypes.Text,
      value: '!'
    }
  ])
})

test('hi % {name}!', () => {
  const text = `hi % {name}!`
  const parser = createParser({ onError: spy })
  const ast = parser.parse(text)

  expect(ast).toMatchSnapshot()
  expect(spy).not.toHaveBeenCalled()

  expect(ast.type).toEqual(NodeTypes.Resource)
  expect(ast.body.type).toEqual(NodeTypes.Message)
  const message = ast.body as MessageNode
  expect(message.items).toHaveLength(3)
  expect(message.items).toMatchObject([
    {
      type: NodeTypes.Text,
      value: 'hi % '
    },
    {
      type: NodeTypes.Named,
      key: 'name'
    },
    {
      type: NodeTypes.Text,
      value: '!'
    }
  ])
})

test('hi %@:name !', () => {
  const text = `hi %@:name !`
  const parser = createParser({ onError: spy })
  const ast = parser.parse(text)

  expect(ast).toMatchSnapshot()
  expect(spy).not.toHaveBeenCalled()

  expect(ast.type).toEqual(NodeTypes.Resource)
  expect(ast.body.type).toEqual(NodeTypes.Message)
  const message = ast.body as MessageNode
  expect(message.items).toHaveLength(3)
  expect(message.items).toMatchObject([
    {
      type: NodeTypes.Text,
      value: 'hi %'
    },
    {
      type: NodeTypes.Linked,
      key: {
        type: NodeTypes.LinkedKey,
        value: 'name'
      }
    },
    {
      type: NodeTypes.Text,
      value: ' !'
    }
  ])
})

test('hi %d %s !', () => {
  const text = `hi %d %s !`
  const parser = createParser({ onError: spy })
  const ast = parser.parse(text)

  expect(ast).toMatchSnapshot()
  expect(spy).not.toHaveBeenCalled()

  expect(ast.type).toEqual(NodeTypes.Resource)
  expect(ast.body.type).toEqual(NodeTypes.Message)
  const message = ast.body as MessageNode
  expect(message.items).toHaveLength(1)
  expect(message.items).toMatchObject([
    {
      type: NodeTypes.Text,
      value: 'hi %d %s !'
    }
  ])
})

test('hi %% name !', () => {
  const text = `hi %% name !`
  const parser = createParser({ onError: spy })
  const ast = parser.parse(text)

  expect(ast).toMatchSnapshot()
  expect(spy).not.toHaveBeenCalled()

  expect(ast.type).toEqual(NodeTypes.Resource)
  expect(ast.body.type).toEqual(NodeTypes.Message)
  const message = ast.body as MessageNode
  expect(message.items).toHaveLength(1)
  expect(message.items).toMatchObject([
    {
      type: NodeTypes.Text,
      value: 'hi %% name !'
    }
  ])
})

test('no apples %| one apple % |  too much apples  ', () => {
  const text = `no apples %| one apple % |  too much apples  `
  const parser = createParser({ onError: spy })
  const ast = parser.parse(text)

  expect(ast).toMatchSnapshot()
  expect(spy).not.toHaveBeenCalled()

  expect(ast.type).toEqual(NodeTypes.Resource)
  expect(ast.body.type).toEqual(NodeTypes.Plural)
  const plural = ast.body as PluralNode
  expect(plural.cases).toHaveLength(3)
  expect(plural.cases).toMatchObject([
    {
      type: NodeTypes.Message,
      items: [
        {
          type: NodeTypes.Text,
          value: 'no apples %'
        }
      ]
    },
    {
      type: NodeTypes.Message,
      items: [
        {
          type: NodeTypes.Text,
          value: 'one apple %'
        }
      ]
    },
    {
      type: NodeTypes.Message,
      items: [
        {
          type: NodeTypes.Text,
          value: 'too much apples  '
        }
      ]
    }
  ])
})
