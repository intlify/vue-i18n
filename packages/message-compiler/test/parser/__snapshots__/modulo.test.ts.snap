// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`%{nickname} %{action} issue %{code} 1`] = `
{
  "body": {
    "end": 35,
    "items": [
      {
        "end": 11,
        "key": "nickname",
        "loc": {
          "end": {
            "column": 12,
            "line": 1,
            "offset": 11,
          },
          "start": {
            "column": 2,
            "line": 1,
            "offset": 1,
          },
        },
        "modulo": true,
        "start": 1,
        "type": 4,
      },
      {
        "end": 12,
        "loc": {
          "end": {
            "column": 13,
            "line": 1,
            "offset": 12,
          },
          "start": {
            "column": 12,
            "line": 1,
            "offset": 11,
          },
        },
        "start": 11,
        "type": 3,
        "value": " ",
      },
      {
        "end": 21,
        "key": "action",
        "loc": {
          "end": {
            "column": 22,
            "line": 1,
            "offset": 21,
          },
          "start": {
            "column": 14,
            "line": 1,
            "offset": 13,
          },
        },
        "modulo": true,
        "start": 13,
        "type": 4,
      },
      {
        "end": 28,
        "loc": {
          "end": {
            "column": 29,
            "line": 1,
            "offset": 28,
          },
          "start": {
            "column": 22,
            "line": 1,
            "offset": 21,
          },
        },
        "start": 21,
        "type": 3,
        "value": " issue ",
      },
      {
        "end": 35,
        "key": "code",
        "loc": {
          "end": {
            "column": 36,
            "line": 1,
            "offset": 35,
          },
          "start": {
            "column": 30,
            "line": 1,
            "offset": 29,
          },
        },
        "modulo": true,
        "start": 29,
        "type": 4,
      },
    ],
    "loc": {
      "end": {
        "column": 36,
        "line": 1,
        "offset": 35,
      },
      "start": {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 35,
  "loc": {
    "end": {
      "column": 36,
      "line": 1,
      "offset": 35,
    },
    "source": "%{nickname} %{action} issue %{code}",
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`hi % {name}! 1`] = `
{
  "body": {
    "end": 12,
    "items": [
      {
        "end": 5,
        "loc": {
          "end": {
            "column": 6,
            "line": 1,
            "offset": 5,
          },
          "start": {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hi % ",
      },
      {
        "end": 11,
        "key": "name",
        "loc": {
          "end": {
            "column": 12,
            "line": 1,
            "offset": 11,
          },
          "start": {
            "column": 6,
            "line": 1,
            "offset": 5,
          },
        },
        "start": 5,
        "type": 4,
      },
      {
        "end": 12,
        "loc": {
          "end": {
            "column": 13,
            "line": 1,
            "offset": 12,
          },
          "start": {
            "column": 12,
            "line": 1,
            "offset": 11,
          },
        },
        "start": 11,
        "type": 3,
        "value": "!",
      },
    ],
    "loc": {
      "end": {
        "column": 13,
        "line": 1,
        "offset": 12,
      },
      "start": {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 12,
  "loc": {
    "end": {
      "column": 13,
      "line": 1,
      "offset": 12,
    },
    "source": "hi % {name}!",
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`hi %% name ! 1`] = `
{
  "body": {
    "end": 12,
    "items": [
      {
        "end": 12,
        "loc": {
          "end": {
            "column": 13,
            "line": 1,
            "offset": 12,
          },
          "start": {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hi %% name !",
      },
    ],
    "loc": {
      "end": {
        "column": 13,
        "line": 1,
        "offset": 12,
      },
      "start": {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 12,
  "loc": {
    "end": {
      "column": 13,
      "line": 1,
      "offset": 12,
    },
    "source": "hi %% name !",
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`hi %@:name ! 1`] = `
{
  "body": {
    "end": 12,
    "items": [
      {
        "end": 4,
        "loc": {
          "end": {
            "column": 5,
            "line": 1,
            "offset": 4,
          },
          "start": {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hi %",
      },
      {
        "end": 10,
        "key": {
          "end": 10,
          "loc": {
            "end": {
              "column": 11,
              "line": 1,
              "offset": 10,
            },
            "start": {
              "column": 7,
              "line": 1,
              "offset": 6,
            },
          },
          "start": 6,
          "type": 7,
          "value": "name",
        },
        "loc": {
          "end": {
            "column": 11,
            "line": 1,
            "offset": 10,
          },
          "start": {
            "column": 5,
            "line": 1,
            "offset": 4,
          },
        },
        "start": 4,
        "type": 6,
      },
      {
        "end": 12,
        "loc": {
          "end": {
            "column": 13,
            "line": 1,
            "offset": 12,
          },
          "start": {
            "column": 11,
            "line": 1,
            "offset": 10,
          },
        },
        "start": 10,
        "type": 3,
        "value": " !",
      },
    ],
    "loc": {
      "end": {
        "column": 13,
        "line": 1,
        "offset": 12,
      },
      "start": {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 12,
  "loc": {
    "end": {
      "column": 13,
      "line": 1,
      "offset": 12,
    },
    "source": "hi %@:name !",
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`hi %{name}! 1`] = `
{
  "body": {
    "end": 11,
    "items": [
      {
        "end": 3,
        "loc": {
          "end": {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
          "start": {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hi ",
      },
      {
        "end": 10,
        "key": "name",
        "loc": {
          "end": {
            "column": 11,
            "line": 1,
            "offset": 10,
          },
          "start": {
            "column": 5,
            "line": 1,
            "offset": 4,
          },
        },
        "modulo": true,
        "start": 4,
        "type": 4,
      },
      {
        "end": 11,
        "loc": {
          "end": {
            "column": 12,
            "line": 1,
            "offset": 11,
          },
          "start": {
            "column": 11,
            "line": 1,
            "offset": 10,
          },
        },
        "start": 10,
        "type": 3,
        "value": "!",
      },
    ],
    "loc": {
      "end": {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "start": {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 11,
  "loc": {
    "end": {
      "column": 12,
      "line": 1,
      "offset": 11,
    },
    "source": "hi %{name}!",
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`hi %d %s ! 1`] = `
{
  "body": {
    "end": 10,
    "items": [
      {
        "end": 10,
        "loc": {
          "end": {
            "column": 11,
            "line": 1,
            "offset": 10,
          },
          "start": {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hi %d %s !",
      },
    ],
    "loc": {
      "end": {
        "column": 11,
        "line": 1,
        "offset": 10,
      },
      "start": {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 10,
  "loc": {
    "end": {
      "column": 11,
      "line": 1,
      "offset": 10,
    },
    "source": "hi %d %s !",
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`no apples %| one apple % |  too much apples   1`] = `
{
  "body": {
    "cases": [
      {
        "end": 0,
        "items": [
          {
            "end": 11,
            "loc": {
              "end": {
                "column": 12,
                "line": 1,
                "offset": 11,
              },
              "start": {
                "column": 1,
                "line": 1,
                "offset": 0,
              },
            },
            "start": 0,
            "type": 3,
            "value": "no apples %",
          },
        ],
        "loc": {
          "end": {
            "column": 12,
            "line": 1,
            "offset": 11,
          },
          "start": {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 2,
      },
      {
        "end": 13,
        "items": [
          {
            "end": 24,
            "loc": {
              "end": {
                "column": 25,
                "line": 1,
                "offset": 24,
              },
              "start": {
                "column": 14,
                "line": 1,
                "offset": 13,
              },
            },
            "start": 13,
            "type": 3,
            "value": "one apple %",
          },
        ],
        "loc": {
          "end": {
            "column": 25,
            "line": 1,
            "offset": 24,
          },
          "start": {
            "column": 14,
            "line": 1,
            "offset": 13,
          },
        },
        "start": 13,
        "type": 2,
      },
      {
        "end": 45,
        "items": [
          {
            "end": 45,
            "loc": {
              "end": {
                "column": 46,
                "line": 1,
                "offset": 45,
              },
              "start": {
                "column": 29,
                "line": 1,
                "offset": 28,
              },
            },
            "start": 28,
            "type": 3,
            "value": "too much apples  ",
          },
        ],
        "loc": {
          "end": {
            "column": 46,
            "line": 1,
            "offset": 45,
          },
          "start": {
            "column": 29,
            "line": 1,
            "offset": 28,
          },
        },
        "start": 28,
        "type": 2,
      },
    ],
    "end": 45,
    "loc": {
      "end": {
        "column": 46,
        "line": 1,
        "offset": 45,
      },
      "start": {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 1,
  },
  "end": 45,
  "loc": {
    "end": {
      "column": 46,
      "line": 1,
      "offset": 45,
    },
    "source": "no apples %| one apple % |  too much apples  ",
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;
